common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/BlockStoreClient.java

```diff
@@ -190,6 +217,52 @@ public void onFailure(Throwable t) {
     }
   }
 
+  private <T> void retry(
+      final int retryCountValue,
+      final int saslRetryCountValue,
+      final int maxRetries,
+      final boolean enableSaslRetries,
+      int delayMs,
+      Supplier<CompletableFuture<T>> action,
+      CompletableFuture<T> future) {
+    action.get()
+            .thenAccept(future::complete)
+            .exceptionally(e -> {
+              int retryCount = retryCountValue;
+              int saslRetryCount = saslRetryCountValue;
+              boolean isIOException = e instanceof IOException
+                      || (e.getCause() != null && e.getCause() instanceof IOException);
+              boolean isSaslTimeout = enableSaslRetries && e instanceof SaslTimeoutException;
+              if (!isSaslTimeout && saslRetryCount > 0) {
+                Preconditions.checkState(retryCount >= saslRetryCount,
+                        "retryCount must be greater than or equal to saslRetryCount");
+                retryCount -= saslRetryCount;
+                saslRetryCount = 0;
+              }
+              boolean hasRemainingRetries = retryCount < maxRetries;
+              boolean shouldRetry = (isSaslTimeout || isIOException) &&
+                      hasRemainingRetries;
+              if (!shouldRetry) {
+                future.completeExceptionally(e);
+              } else {
+                if (enableSaslRetries && e instanceof SaslTimeoutException) {
+                  saslRetryCount += 1;
+                }
+                retryCount += 1;
+                int finalRetryCount = retryCount;
+                int finalSaslRetryCount = saslRetryCount;
```

Comment:
super nit:
```suggestion
                final int finalRetryCount = retryCount;
                final int finalSaslRetryCount = saslRetryCount;
```


This does get inferred to be `final` - so no longer strictly necessary - but more aligned with the intent and the var name